barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_alu_trace.cpp

```diff
@@ -963,4 +964,105 @@ FF AvmAluTraceBuilder::op_shl(FF const& a, FF const& b, AvmMemoryTag in_tag, uin
+            .remainder = a,
+
+        }));
+        return 0;
+    }
+    // Decompose a and primality check that b*c < p when a is a 256-bit integer
+    auto [a_lo, a_hi] = decompose(b_u256 * c_u256, 128);
+    auto [p_sub_a_lo, p_sub_a_hi, p_a_borrow] = gt_witness(FF::modulus, b_u256 * c_u256);
+    // Decompose the divisor
+    auto [divisor_lo, divisor_hi] = decompose(b_u256, 64);
+    // Decompose the quotient
+    auto [quotient_lo, quotient_hi] = decompose(c_u256, 64);
+    uint256_t partial_prod = divisor_lo * quotient_hi + divisor_hi * quotient_lo;
+    // Decompose the partial product
+    auto [partial_prod_lo, partial_prod_hi] = decompose(partial_prod, 64);
+
+    FF b_hi = b_u256 - rem_u256 - 1;
+
+    // 64 bit range checks for the divisor and quotient limbs
+    // Spread over two rows
+    std::array<uint16_t, 8> div_u64_rng_chk;
+    std::array<uint16_t, 8> div_u64_rng_chk_shifted;
+    for (size_t i = 0; i < 4; i++) {
+        div_u64_rng_chk.at(i) = uint16_t(divisor_lo >> (16 * i));
+        div_u64_rng_chk.at(i + 4) = uint16_t(divisor_hi >> (16 * i));
+        div_u64_range_chk_counters[i][uint16_t(divisor_lo >> (16 * i))]++;
+        div_u64_range_chk_counters[i + 4][uint16_t(divisor_hi >> (16 * i))]++;
+
+        div_u64_rng_chk_shifted.at(i) = uint16_t(quotient_lo >> (16 * i));
+        div_u64_rng_chk_shifted.at(i + 4) = uint16_t(quotient_hi >> (16 * i));
+        div_u64_range_chk_counters[i][uint16_t(quotient_lo >> (16 * i))]++;
+        div_u64_range_chk_counters[i + 4][uint16_t(quotient_hi >> (16 * i))]++;
+    }
+
+    // Each hi and lo limb is range checked over 128 bits
+    // Load the range check values into the ALU registers
+    auto hi_lo_limbs = std::vector<uint256_t>{ a_lo, a_hi, partial_prod, b_hi, p_sub_a_lo, p_sub_a_hi };
+    AvmAluTraceBuilder::AluTraceEntry row{
+        .alu_clk = clk,
+        .alu_op_div = true,
+        .alu_ff_tag = in_tag == AvmMemoryTag::FF,
+        .alu_u8_tag = in_tag == AvmMemoryTag::U8,
+        .alu_u16_tag = in_tag == AvmMemoryTag::U16,
+        .alu_u32_tag = in_tag == AvmMemoryTag::U32,
+        .alu_u64_tag = in_tag == AvmMemoryTag::U64,
+        .alu_u128_tag = in_tag == AvmMemoryTag::U128,
+        .alu_ia = a,
+        .alu_ib = b,
+        .alu_ic = FF{ c_u256 },
+        .remainder = rem_u256,
+        .divisor_lo = divisor_lo,
+        .divisor_hi = divisor_hi,
+        .quotient_lo = quotient_lo,
+        .quotient_hi = quotient_hi,
+        .partial_prod_lo = partial_prod_lo,
+        .partial_prod_hi = partial_prod_hi,
+        .div_u64_range_chk_sel = true,
+        .div_u64_range_chk = div_u64_rng_chk,
+
+    };
+    // We perform the range checks here
+    std::vector<AvmAluTraceBuilder::AluTraceEntry> rows = cmp_range_check_helper(row, hi_lo_limbs);
+    alu_trace.insert(alu_trace.end(), rows.begin(), rows.end());
+    // Add the range checks for the quotient limbs in the next row
+    alu_trace.at(1).div_u64_range_chk = div_u64_rng_chk_shifted;
```

Comment:
Is the absolute index == 1 correct ? I do not think so if alu is handling several operations.
Same remark for the next line.

Reply:
Good catch!